File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-field-type/src/main/resources/META-INF/resources/ObjectField/ObjectField.js
/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {useResource} from '@clayui/data-provider';
import {usePrevious} from '@liferay/frontend-js-react-web';
import {
	getFields,
	getObjectFieldName,
	getSelectedValue,
	useFormState,
} from 'data-engine-js-components-web';
import {fetch} from 'frontend-js-web';
import React, {useEffect, useMemo} from 'react';

import Select from '../Select/Select.es';

const dataTypes = {
	double: ['double', 'bigdecimal'],
	image: ['blob'],
	integer: ['integer', 'long'],
};

const normalizeDataType = (type) => {
	const formattedType = type.toLowerCase();

	return dataTypes[formattedType] ?? formattedType;
};

const ObjectField = ({
	label,
	objectFields,
	onChange,
	readOnly,
	spritemap,
	value = {},
	visible,
}) => {
	const {
		formBuilder: {
			focusedField: {dataType, type: focusedFieldType},
			pages,
		},
	} = useFormState();

	const normalizedDataType = useMemo(() => normalizeDataType(dataType), [
		dataType,
	]);

	const options = useMemo(() => {
		const filteredObjectFields = objectFields.filter(
			({listTypeDefinitionId, relationshipType, type}) => {
				if (
					!listTypeDefinitionId &&
					(focusedFieldType === 'radio' ||
						focusedFieldType === 'select') &&
					normalizedDataType.includes(type.toLowerCase())
				) {
					return false;
				}
				else if (
					listTypeDefinitionId &&
					(focusedFieldType === 'checkbox_multiple' ||
						focusedFieldType === 'color' ||
						focusedFieldType === 'grid' ||
						focusedFieldType === 'rich_text' ||
						focusedFieldType === 'text') &&
					normalizedDataType.includes(type.toLowerCase())
				) {
					return false;
				}
				else if (focusedFieldType === 'text' && type === 'Clob') {
					return true;
				}
				else if (relationshipType) {
					return false;
				}

				return normalizedDataType.includes(type.toLowerCase());
			}
		);

		if (filteredObjectFields.length) {
			const mappedFields = getFields(pages)
				.map((field) => {
					const objectFieldName = getObjectFieldName(field);

					return (
						objectFieldName &&
						getSelectedValue(objectFieldName.value)
					);
				})
				.filter(Boolean);

			return filteredObjectFields.map(({label, name}) => ({
				disabled: !!mappedFields.includes(name),
				label: label[themeDisplay.getDefaultLanguageId()] ?? name,
				value: name,
			}));
		}
		else {
			const emptyStateMessage = Liferay.Language.get(
				'there-are-no-compatible-object-fields-to-map'
			);

			return [
				{
					disabled: true,
					label: emptyStateMessage,
					value: emptyStateMessage,
				},
			];
		}
	}, [focusedFieldType, normalizedDataType, objectFields, pages]);

	return (
		<Select
			label={label}
			name="selectedObjectField"
			onChange={onChange}
			options={options}
			readOnly={readOnly}
			showEmptyOption={!!options.length}
			spritemap={spritemap}
			value={getSelectedValue(value)}
			visible={visible}
		/>
	);
};

const ObjectDefinitionObjectField = ({
	label,
	objectDefinitionId,
	onChange,
	readOnly,
	spritemap,
	value = {},
	visible,
}) => {
	const {refetch, resource} = useResource({
		fetch,
		fetchPolicy: 'cache-first',
		link: `${window.location.origin}/o/object-admin/v1.0/object-definitions/${objectDefinitionId}`,
	});

	const previousObjectDefinitionId = usePrevious(objectDefinitionId);

	useEffect(() => {
		if (
			objectDefinitionId &&
			objectDefinitionId !== previousObjectDefinitionId
		) {
			refetch();
		}
	}, [objectDefinitionId, previousObjectDefinitionId, refetch]);

	const options =
		resource?.objectFields?.map(({label, name}) => {
			return {
				label: label[themeDisplay.getDefaultLanguageId()] ?? name,
				value: name,
			};
		}) || [];

	return (
		<Select
			label={label}
			name="selectedObjectField"
			onChange={onChange}
			options={options}
			readOnly={readOnly}
			showEmptyOption={!!options.length}
			spritemap={spritemap}
			value={value}
			visible={visible}
		/>
	);
};

const ObjectFieldWrapper = ({objectDefinitionId, ...props}) => {
	const {objectFields} = useFormState();

	if (!objectFields?.length) {
		if (objectDefinitionId) {
			return (
				<ObjectDefinitionObjectField
					objectDefinitionId={objectDefinitionId}
					{...props}
				/>
			);
		}

		return null;
	}

	return <ObjectField objectFields={objectFields} {...props} />;
};

export default ObjectFieldWrapper;

File: modules/apps/object/object-admin-rest-client/src/main/java/com/liferay/object/admin/rest/client/dto/v1_0/ObjectField.java
/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

package com.liferay.object.admin.rest.client.dto.v1_0;

import com.liferay.object.admin.rest.client.function.UnsafeSupplier;
import com.liferay.object.admin.rest.client.serdes.v1_0.ObjectFieldSerDes;

import java.io.Serializable;

import java.util.Map;
import java.util.Objects;

import javax.annotation.Generated;

/**
 * @author Javier Gamarra
 * @generated
 */
@Generated("")
public class ObjectField implements Cloneable, Serializable {

	public static ObjectField toDTO(String json) {
		return ObjectFieldSerDes.toDTO(json);
	}

	public DBType getDBType() {
		return DBType;
	}

	public String getDBTypeAsString() {
		if (DBType == null) {
			return null;
		}

		return DBType.toString();
	}

	public void setDBType(DBType DBType) {
		this.DBType = DBType;
	}

	public void setDBType(
		UnsafeSupplier<DBType, Exception> DBTypeUnsafeSupplier) {

		try {
			DBType = DBTypeUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected DBType DBType;

	public Map<String, Map<String, String>> getActions() {
		return actions;
	}

	public void setActions(Map<String, Map<String, String>> actions) {
		this.actions = actions;
	}

	public void setActions(
		UnsafeSupplier<Map<String, Map<String, String>>, Exception>
			actionsUnsafeSupplier) {

		try {
			actions = actionsUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected Map<String, Map<String, String>> actions;

	public BusinessType getBusinessType() {
		return businessType;
	}

	public String getBusinessTypeAsString() {
		if (businessType == null) {
			return null;
		}

		return businessType.toString();
	}

	public void setBusinessType(BusinessType businessType) {
		this.businessType = businessType;
	}

	public void setBusinessType(
		UnsafeSupplier<BusinessType, Exception> businessTypeUnsafeSupplier) {

		try {
			businessType = businessTypeUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected BusinessType businessType;

	public Long getId() {
		return id;
	}

	public void setId(Long id) {
		this.id = id;
	}

	public void setId(UnsafeSupplier<Long, Exception> idUnsafeSupplier) {
		try {
			id = idUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected Long id;

	public Boolean getIndexed() {
		return indexed;
	}

	public void setIndexed(Boolean indexed) {
		this.indexed = indexed;
	}

	public void setIndexed(
		UnsafeSupplier<Boolean, Exception> indexedUnsafeSupplier) {

		try {
			indexed = indexedUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected Boolean indexed;

	public Boolean getIndexedAsKeyword() {
		return indexedAsKeyword;
	}

	public void setIndexedAsKeyword(Boolean indexedAsKeyword) {
		this.indexedAsKeyword = indexedAsKeyword;
	}

	public void setIndexedAsKeyword(
		UnsafeSupplier<Boolean, Exception> indexedAsKeywordUnsafeSupplier) {

		try {
			indexedAsKeyword = indexedAsKeywordUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected Boolean indexedAsKeyword;

	public String getIndexedLanguageId() {
		return indexedLanguageId;
	}

	public void setIndexedLanguageId(String indexedLanguageId) {
		this.indexedLanguageId = indexedLanguageId;
	}

	public void setIndexedLanguageId(
		UnsafeSupplier<String, Exception> indexedLanguageIdUnsafeSupplier) {

		try {
			indexedLanguageId = indexedLanguageIdUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected String indexedLanguageId;

	public Map<String, String> getLabel() {
		return label;
	}

	public void setLabel(Map<String, String> label) {
		this.label = label;
	}

	public void setLabel(
		UnsafeSupplier<Map<String, String>, Exception> labelUnsafeSupplier) {

		try {
			label = labelUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected Map<String, String> label;

	public Long getListTypeDefinitionId() {
		return listTypeDefinitionId;
	}

	public void setListTypeDefinitionId(Long listTypeDefinitionId) {
		this.listTypeDefinitionId = listTypeDefinitionId;
	}

	public void setListTypeDefinitionId(
		UnsafeSupplier<Long, Exception> listTypeDefinitionIdUnsafeSupplier) {

		try {
			listTypeDefinitionId = listTypeDefinitionIdUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected Long listTypeDefinitionId;

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public void setName(UnsafeSupplier<String, Exception> nameUnsafeSupplier) {
		try {
			name = nameUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected String name;

	public RelationshipType getRelationshipType() {
		return relationshipType;
	}

	public String getRelationshipTypeAsString() {
		if (relationshipType == null) {
			return null;
		}

		return relationshipType.toString();
	}

	public void setRelationshipType(RelationshipType relationshipType) {
		this.relationshipType = relationshipType;
	}

	public void setRelationshipType(
		UnsafeSupplier<RelationshipType, Exception>
			relationshipTypeUnsafeSupplier) {

		try {
			relationshipType = relationshipTypeUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected RelationshipType relationshipType;

	public Boolean getRequired() {
		return required;
	}

	public void setRequired(Boolean required) {
		this.required = required;
	}

	public void setRequired(
		UnsafeSupplier<Boolean, Exception> requiredUnsafeSupplier) {

		try {
			required = requiredUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected Boolean required;

	public Type getType() {
		return type;
	}

	public String getTypeAsString() {
		if (type == null) {
			return null;
		}

		return type.toString();
	}

	public void setType(Type type) {
		this.type = type;
	}

	public void setType(UnsafeSupplier<Type, Exception> typeUnsafeSupplier) {
		try {
			type = typeUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected Type type;

	@Override
	public ObjectField clone() throws CloneNotSupportedException {
		return (ObjectField)super.clone();
	}

	@Override
	public boolean equals(Object object) {
		if (this == object) {
			return true;
		}

		if (!(object instanceof ObjectField)) {
			return false;
		}

		ObjectField objectField = (ObjectField)object;

		return Objects.equals(toString(), objectField.toString());
	}

	@Override
	public int hashCode() {
		String string = toString();

		return string.hashCode();
	}

	public String toString() {
		return ObjectFieldSerDes.toJSON(this);
	}

	public static enum BusinessType {

		BOOLEAN("Boolean"), DATE("Date"), DECIMAL("Decimal"),
		INTEGER("Integer"), LONG_INTEGER("LongInteger"), LONG_TEXT("LongText"),
		PICKLIST("Picklist"), PRECISION_DECIMAL("PrecisionDecimal"),
		RELATIONSHIP("Relationship"), TEXT("Text");

		public static BusinessType create(String value) {
			for (BusinessType businessType : values()) {
				if (Objects.equals(businessType.getValue(), value) ||
					Objects.equals(businessType.name(), value)) {

					return businessType;
				}
			}

			return null;
		}

		public String getValue() {
			return _value;
		}

		@Override
		public String toString() {
			return _value;
		}

		private BusinessType(String value) {
			_value = value;
		}

		private final String _value;

	}

	public static enum DBType {

		BIG_DECIMAL("BigDecimal"), BOOLEAN("Boolean"), CLOB("Clob"),
		DATE("Date"), DOUBLE("Double"), INTEGER("Integer"), LONG("Long"),
		STRING("String");

		public static DBType create(String value) {
			for (DBType dbType : values()) {
				if (Objects.equals(dbType.getValue(), value) ||
					Objects.equals(dbType.name(), value)) {

					return dbType;
				}
			}

			return null;
		}

		public String getValue() {
			return _value;
		}

		@Override
		public String toString() {
			return _value;
		}

		private DBType(String value) {
			_value = value;
		}

		private final String _value;

	}

	public static enum RelationshipType {

		ONE_TO_MANY("oneToMany"), ONE_TO_ONE("oneToOne");

		public static RelationshipType create(String value) {
			for (RelationshipType relationshipType : values()) {
				if (Objects.equals(relationshipType.getValue(), value) ||
					Objects.equals(relationshipType.name(), value)) {

					return relationshipType;
				}
			}

			return null;
		}

		public String getValue() {
			return _value;
		}

		@Override
		public String toString() {
			return _value;
		}

		private RelationshipType(String value) {
			_value = value;
		}

		private final String _value;

	}

	public static enum Type {

		BIG_DECIMAL("BigDecimal"), BOOLEAN("Boolean"), CLOB("Clob"),
		DATE("Date"), DOUBLE("Double"), INTEGER("Integer"), LONG("Long"),
		STRING("String");

		public static Type create(String value) {
			for (Type type : values()) {
				if (Objects.equals(type.getValue(), value) ||
					Objects.equals(type.name(), value)) {

					return type;
				}
			}

			return null;
		}

		public String getValue() {
			return _value;
		}

		@Override
		public String toString() {
			return _value;
		}

		private Type(String value) {
			_value = value;
		}

		private final String _value;

	}

}